### ISSUE TITLE
On the : command line, "Esc" and ^C should cancel the command

### ISSUE THREAD

**Issue by martinwguy** [martinwguy/xvi#19]
Neither does anything at present: the only ways out are by deleting the whole line with backspace or pressing [Enter] to execute it.

**Comment by jvvv** [comment]
"ESC" is supposed to terminate the command mode and discard input, not execute it:
http://pubs.opengroup.org/onlinepubs/9699919799/utilities/vi.html#tag_20_152_13_16

**Comment by martinwguy** [comment]
In xvi, ESC does partial filename completion. If you go into the src directory and go "xvi[Enter]:e cmdl[Esc] it fills in the "ine." for you if there are cmdline-c and cmdline.o.
vim has something similar when you press Tab after commands taking filename args, but cycling through the candidate filenames if there are more than one; xvi does it anywhere on any command line because the command is not parsed until the line is complete.

I'd tend to drop the filename completion and make ESC cancel the input line instead. What do you think?

Control-C is not so easy as it doesn't arrive a a char, but causes a signal instead, which is caught but does *not* cause the keyboard-reading "read()" system call to return early.

**Comment by jvvv** [comment]
Mainly I was distraught to think that the ESC key would be used for execution of command. Having ESC cancel makes sense to me. As for filename completion, TAB does seem to be common from my experience, so might be more intuitive if you were to keep the feature. As far as catching ^C, wouldn't it be possible to have the signal handler set a global variable that could be checked during cmdmode input and in turn would cancel input when the variable is set? The check could probably be done in cmdmode.c:cmd_input() before the first if() statement at line 117. Just an idea.

**Comment by martinwguy** [comment]
On 09/11/2016, jvvv <notifications@github.com> wrote:
> Mainly I was distraught to think that the ESC key would be used for
> execution of command. Having ESC cancel makes sense to me.

agree.

> As for filename completion, TAB does seem to be common, so might be more
> intuitive if you were to keep the feature.

My impulse was to dump the completion but only cos I don't use it but
I guess it's more faithful to keep the original design with less
surprising keys.

> As far as catching ^C, wouldn't it be possible to have the
> signal handler set a global variable

It already does that. The variable, kbdintr, is set but itt doesn't
interrupt the read() system call that is waiting on the keyboard/ at
least on Linux and BSD (on AIX it does). Maybe checking the machine
state in the irq handler and doing a longjmp() if it's in the command
line might work.

**Comment by jvvv** [comment]
Sorry about missing that variable. Had I read deeper into the code, the SIGINT handler would have been obvious. I am slowly digging into the source, but will take a some while yet to get a firmer footing. My gut feeling is that if an editor requires looking at anything on the irq handler level of the system, then that is getting too machine specific for application code... just my 2 cents. Perhaps ^C handling in the ex cmdline could be postponed for later.

**Comment by jvvv** [comment]
I'm wondering why read() is blocking when raw/cbreak mode has supposedly been setup.

**Comment by jvvv** [comment]
I have had some time to start getting deeper into xvi code. I have a pull request https://github.com/martinwguy/xvi/pull/31 for the ^C part of the issue.

**Comment by martinwguy** [comment]
Thanks! The solution of poking "ESC" into the input buffer in the
interrupt routine is elegant. Unfortunately the "FIONREAD" change, to
avoid "read" blocking, makes xvi consume 100% CPU all the time on
Unix. Is there another way to solve this?

**Comment by jvvv** [comment]
I will take another stab at it tomorrow afternoon. I should have thought to check top for the cpu usage, especially when making that ioctl addition. It slipped my mind.

**Comment by jvvv** [comment]
Is there anything left to cover in order to close this issue? I was under the impression that the last pr for dealing with ^C and ESC put us in line with what was desired, but I may be mistaken. Let me know if there are other use cases, or ones that are not completely dealt with, and I can address them.

**Comment by martinwguy** [comment]
A minor extension: to print "Interrupted" when a double-character command is abandoned with ^C.

Also, sometimes the command line takes five seconds to interrupt but it seems random.
:sdfjhskfjh[Control-C] seems a good candidate - copy that sequence try it half a dozen times - sometimes there is a 5 second delay between pressing ^C and it printing "Interrupted" and returning the cursor to the display area.

**Comment by jvvv** [comment]
I've notice delays in with interrupt reception as well. I'm still looking into it.

**Comment by jvvv** [comment]
I've added two commits to #38 that attempt to prevent delays in ^C handling. Seems pretty responsive so far to me.

**Comment by martinwguy** [comment]
Fixed by 4fffa72a

**Comment by martinwguy** [comment]
Looks good.
